common/discovery.go

```diff
@@ -0,0 +1,165 @@
+// Copyright © 2018 Barthelemy Vessemont
+// GNU General Public License version 3
+
+package common
+
+import (
+	"fmt"
+	"github.com/hashicorp/consul/api"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"sort"
+	"strings"
+)
+
+//TODO detect probe stuck
+func UpdateEverKnownNodes(allEverKnownNodes []string, nodes []Node) []string {
+	for _, node := range nodes {
+		// TODO: Replace by a real struct instead of a string concatenation...
+		// also allEverKnownNodes leak memory as it never delete old/cleaned items
+		serializedNode := fmt.Sprintf("%v|%v", node.Name, node.Cluster)
+		if contains(allEverKnownNodes, serializedNode) == false {
+			allEverKnownNodes = append(allEverKnownNodes, serializedNode)
+		}
+	}
+	sort.Strings(allEverKnownNodes)
+	return allEverKnownNodes
+}
+
+func NewClient(consulTarget string) (*api.Client, error) {
+	consulConfig := api.DefaultConfig()
+	consulConfig.Address = consulTarget
+	consul, err := api.NewClient(consulConfig)
+	if err != nil {
+		ErrorsCount.Inc()
+		return nil, errors.Wrapf(err, "Failed to create consul client with target %s", consulTarget)
+	}
+	return consul, nil
+}
+
+func DiscoverNodesForService(consul *api.Client, serviceName string) ([]Node, error) {
+	catalogServices, _, err := consul.Catalog().Service(
+		serviceName, "",
+		&api.QueryOptions{AllowStale: true, RequireConsistent: false, UseCache: true},
+	)
+	if err != nil {
+		log.Error("Consul Discovery failed: ", err.Error())
+		ErrorsCount.Inc()
+		return nil, err
+	}
+
+	var nodeList []Node
+	for _, svc := range catalogServices {
+		var addr = svc.Address
+		if svc.ServiceAddress != "" {
+			addr = svc.ServiceAddress
+		}
+
+		log.Debug("Service discovered: ", svc.Node, " (", addr, ":", svc.ServicePort, ")")
+		nodeList = append(nodeList, Node{
+			Name:    svc.Node,
+			Ip:      addr,
+			Port:    svc.ServicePort,
+			Scheme:  schemeFromTags(svc.ServiceTags),
+			Cluster: valueFromTags("cluster_name", svc.ServiceTags),
+		})
+	}
+
+	nodesCount := len(nodeList)
+	log.Debug(nodesCount, " nodes found")
+
+	return nodeList, nil
+}
+
+func GetServices(consul *api.Client, consulTag string) (map[string]Cluster, error) {
+	consulServices, _, err := consul.Catalog().Services(nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get services from consul")
+	}
+
+	var services = make(map[string]Cluster)
+	var service Cluster
+	for serviceName := range consulServices {
+		for i := range consulServices[serviceName] {
+			if consulServices[serviceName][i] == consulTag {
+				// Check cluster not already added
+				cluster := valueFromTags("cluster_name", consulServices[serviceName])
+				// TODO ensure we use https when available?
```

Comment:
TODO is outdated, we take the scheme (http/https) at L92

Reply:
Not outdater we take first consul entry but first one can be the http definition then we don't override if we found the https one
Idea is to enforce usage of https